core/src/workflow/machines/upsert_search_attributes_state_machine.rs

```diff
@@ -1,40 +1,239 @@
+    // state when it receives the CommandScheduled event that is a result of looping back the
+    // Command initially packaged with NewMachineWithCommand (see upsert_search_attrs)
+    Created --(CommandScheduled) --> CommandIssued;
 
-    UpsertCommandCreated --(CommandUpsertWorkflowSearchAttributes) --> UpsertCommandCreated;
-    UpsertCommandCreated --(UpsertWorkflowSearchAttributes, on_upsert_workflow_search_attributes) --> UpsertCommandRecorded;
+    // Having sent the command to the server, the machine transitions into a terminal state (Done)
+    // upon observing a history event indicating that the command has been recorded. Note that this
+    // does not imply that the command has been _executed_, only that it _will be_ executed at some
+    // point in the future.
+    CommandIssued --(CommandRecorded) --> Done;
 }
 
-#[derive(thiserror::Error, Debug)]
-pub(super) enum UpsertSearchAttributesMachineError {}
+// Instantiates an UpsertSearchAttributesMachine and packs it together with an initial command
+// to apply the provided search attribute update.
+pub(super) fn upsert_search_attrs(
+    attribs: UpsertWorkflowSearchAttributes,
+) -> NewMachineWithCommand {
+    let mut sm = UpsertSearchAttributesMachine::new();
+    let cmd = Command {
+        command_type: CommandType::UpsertWorkflowSearchAttributes as i32,
+        attributes: Some(attribs.into()),
+    };
+    NewMachineWithCommand {
+        command: cmd,
+        machine: sm.into(),
+    }
+}
 
-pub(super) enum UpsertSearchAttributesCommand {}
+// Unused but must exist
+type SharedState = ();
 
-#[derive(Default, Clone)]
+// The state-machine-specific set of commands that are the results of state transition in the
+// UpsertSearchAttributesMachine.
+#[derive(Debug, derive_more::Display)]
+pub(super) enum UpsertSearchAttributesMachineCommand {
+    #[display(fmt = "SendUpsertCommand")]
+    SendUpsertCommand(Attributes),
+}
+
+// The state of the UpsertSearchAttributesMachine at time zero (i.e. at instantiation)
+#[derive(Debug, Default, Clone, derive_more::Display)]
 pub(super) struct Created {}
 
-impl Created {
-    pub(super) fn on_schedule(
-        self,
-    ) -> UpsertSearchAttributesMachineTransition<UpsertCommandCreated> {
-        unimplemented!()
+// Once the UpsertSearchAttributesMachine has been known to have issued the upsert command to
+// higher-level machinery, it transitions into this state.
+#[derive(Debug, Default, Clone, derive_more::Display)]
+pub(super) struct CommandIssued {}
+
+// Once the server has recorded its receipt of the search attribute update, the
+// UpsertSearchAttributesMachine transitions into this terminal state.
+#[derive(Debug, Default, Clone, derive_more::Display)]
+pub(super) struct Done {}
+
+impl UpsertSearchAttributesMachine {
+    fn new() -> Self {
+        Self {
+            state: Created {}.into(),
+            shared_state: (),
+        }
+    }
+
+    fn state(&self) -> String {
+        self.state.to_string()
     }
```

Comment:
This is auto-implemented as `.state().name()` I believe

Reply:
there's no `.name()` but I have `.to_string()` which works. Thanks